backend/enums/activity.js

```diff
@@ -1,116 +1,107 @@
-      ...sharedPostsActivity,
-    ]
-    activity.sort((a, b) => (a.timestamp < b.timestamp ? 1 : -1))
-
-    return activity
-  }
-
-  static convertToUserActivity(activity, postId, postTime) {
-    return {
-      postID: postId,
-      timestamp: Date.parse(postTime),
-      activity: activity.type,
-    }
-  }
-
-  static async convertToFeedActivity(activity, postId, authorId, postTime) {
-    const post = await models.posts.findByPk(postId)
-    const postDto = await PostDTO.convertToDto(post)
-    return {
-      post: postDto,
-      timestamp: Date.parse(postTime),
-      activity: activity.type,
-      userId: authorId,
-    }
+export default class Activity {
+  constructor(postDto, interactingUser, activityType, activityTime) {
+    // Fields named for compatability with old usage
+    this.postID = postDto.id
+    this.post = postDto
+    this.userId = interactingUser
+    this.activity = activityType
+    this.timestamp = activityTime
   }
 
-  static async getUnconvertedActivity(userId) {
-    // Retrieve user's posts, including the shared and liked posts
-    const postsDB = await models.posts.findAll({
+  // Create a list of activities by a user in chronological order
+  static async getUserActivities(userId) {
+    let activities = []
+    let ownPosts = await models.posts.findAll({
       where: {
         author: userId,
       },
     })
+    // async forEach didn't seem to await properly
+    for (const post of ownPosts) {
+      activities.push(
+        new Activity(
+          await PostDTO.convertToDto(post),
+          userId,
+          ActivityType.POSTED.type,
+          Date.parse(post.createdAt)
+        )
+      )
+    }
 
-    const sharedPosts = await models.sharedPost.findAll({
+    let sharedPosts = await models.sharedPost.findAll({
       where: {
         userID: userId,
       },
     })
+    for (const sharedPost of sharedPosts) {
+      // Get each real post DTO by querying individually
+      let post = await models.posts.findOne({
+        where: {
+          id: sharedPost.postId,
+        },
+      })
+      activities.push(
+        new Activity(
+          await PostDTO.convertToDto(post),
+          userId,
+          ActivityType.SHARED.type,
+          Date.parse(sharedPost.createdAt)
+        )
+      )
+    }
 
-    const likedPosts = await models.likedPost.findAll({
+    let likedPosts = await models.likedPost.findAll({
       where: {
         userID: userId,
       },
     })
+    for (const likedPost of likedPosts) {
+      let post = await models.posts.findOne({
+        where: {
+          id: likedPost.postId,
+        },
+      })
+      activities.push(
+        new Activity(
+          await PostDTO.convertToDto(post),
+          userId,
+          ActivityType.LIKED.type,
+          Date.parse(likedPost.createdAt)
+        )
+      )
+    }
+
+    activities.sort((a, b) => {
+      return a.timestamp < b.timestamp ? 1 : -1
+    })
+    return activities
+  }
```

Comment:
Can you use the Promise.all() instead. it runs the await concurrently, it is too slow to wait for each await in the current for loop.